Feature/apple 51 improve error feedback when app install fails due to

abbydomi opened the pull request:
This pull request adds comprehensive mocking and testing support for login and update flows, introduces disk space checks before downloads, and enhances configuration handling with a new build size property. The main changes include new mock classes, expanded test coverage for login scenarios, and updates to the SDK to track and persist the last build size and disk space errors.

**Testing and Mocking Enhancements:**
* Added new mock classes for `LoginRepository`, `DownloadService`, and `SessionPersister` to facilitate isolated and robust unit testing of login and update flows. (`AppliveryBehaviorTests/Mocks/LoginRepositoryMock.swift` [[1]](diffhunk://#diff-4ace2fd7eb60d8e0a6fcd2820a7e9c6d7fdc9a5db02d4b9d735119a65d032be5R1-R23) `AppliveryBehaviorTests/Mocks/DownloadServiceMock.swift` [[2]](diffhunk://#diff-4e84350a36be7a6a57ea70a255541d654a3439093a14e0f84d58351a582f61c5R1-R26) `AppliveryBehaviorTests/Mocks/MockSessionPersister.swift` [[3]](diffhunk://#diff-565dded0812fe9e02ab214bd46ac3d99852bf48d8227581920be4d4e633706daR1-R34)
* Extended `AppMock` to allow stubbing available device space and implemented the `deviceAvailableSpace` method, enabling disk space checks in tests. (`AppliveryBehaviorTests/Mocks/AppMock.swift` [[1]](diffhunk://#diff-7b3454026ac61d3e1cfa74ec923fc47ab0a58fbd0c53e671688c6b039bcff0bcR21) [[2]](diffhunk://#diff-7b3454026ac61d3e1cfa74ec923fc47ab0a58fbd0c53e671688c6b039bcff0bcR126-R129)

**Test Coverage Improvements:**
* Added a comprehensive suite of tests for the login service, covering successful login, 401 failures, authorization requests with/without tokens, and download scenarios with sufficient and insufficient disk space. (`AppliveryBehaviorTests/ServicesTests/LoginServiceTests.swift` [AppliveryBehaviorTests/ServicesTests/LoginServiceTests.swiftR1-R266](diffhunk://#diff-a2ca7346cd3c66b4400660e2c60d9cac1c22326eef4de0050c4fcbc594e28b5cR1-R266))

**Configuration and Persistence Updates:**
* Introduced the `lastBuildSize` property to `SDKData` and ensured it is correctly persisted and restored via `ConfigPersister`, as well as included in mock/test data. (`AppliverySDK/Model/Config.swift` [[1]](diffhunk://#diff-e9d84bb2604a3ce004fade7dd0eb68758c5a5e29087f598d8da2ede484f7a839R38) `AppliverySDK/Data/Persisters/ConfigPersister.swift` [[2]](diffhunk://#diff-39278ad768135a476051fc6e26cd5c5cb58f2f78b0e03175b4a69ebeee05fc1fR19) [[3]](diffhunk://#diff-39278ad768135a476051fc6e26cd5c5cb58f2f78b0e03175b4a69ebeee05fc1fL62-R62) [[4]](diffhunk://#diff-39278ad768135a476051fc6e26cd5c5cb58f2f78b0e03175b4a69ebeee05fc1fR77) AppliveryBehaviorTests/MockData/ConfigMockData.swift [[5]](diffhunk://#diff-1b18a67fb58f1f1ccd9d14bbe5ade33a38c12bd5427cf20840fca42a7dd388ebL25-R26) AppliveryBehaviorTests/Mocks/ConfigServiceMock.swift [[6]](diffhunk://#diff-96769c275c7e18bfed6bd9a06966f00567e5da5438ef66b614d413493dd7a577L41-R42) AppliveryBehaviorTests/ServicesTests/UpdateServiceTests.swift [[7]](diffhunk://#diff-f6c9f0be47e3e78b7e1be63629053740a00819d1b289bab2bce006e02617ab9fL131-R132) [[8]](diffhunk://#diff-f6c9f0be47e3e78b7e1be63629053740a00819d1b289bab2bce006e02617ab9fL155-R157)

**SDK and Protocol Improvements:**
* Updated `UpdateError` to conform to `Error` and added new cases for disk space errors, enabling improved error handling and reporting for update/download operations. (`AppliverySDK/Applivery/AppliverySDK.swift` [[1]](diffhunk://#diff-f86d3e567fed0c23b2547250bbe6181e247a7794549efc32abd602e79511da51L44-R44) [[2]](diffhunk://#diff-f86d3e567fed0c23b2547250bbe6181e247a7794549efc32abd602e79511da51R64-R69)
* Introduced `SessionPersisterProtocol` and updated `LoginService` to depend on the protocol rather than the concrete type for better testability. (`AppliverySDK/Data/Persisters/SessionPersister.swift` [[1]](diffhunk://#diff-fff318f6ee5e21ef33442e72c6c464513c5a1e954f3baa6d5c90554558476fcaR21-R28) `AppliverySDK/Services/LoginService.swift` [[2]](diffhunk://#diff-07063ac10a6f5b5974497709cdf7284a1278e26efb098068c40d883b51e824eeL28-R28) [[3]](diffhunk://#diff-07063ac10a6f5b5974497709cdf7284a1278e26efb098068c40d883b51e824eeL43-R43)

abbydomi committed:
feat: add disk space error handling and build size tracking

Introduced new UpdateError cases for disk space issues and updated SDKData and ConfigPersister to store and retrieve the last build size. These changes improve update reliability by allowing the SDK to detect and handle insufficient disk space scenarios.

abbydomi committed:
feat: Add disk space check before download in LoginService

Introduced a check for available device storage before initiating a download in LoginService. Refactored SessionPersister to use a protocol, allowing for dependency injection and improved testability.

abbydomi committed:
feat: Add mocks and tests for LoginService and update config mocks

Added new mock classes (DownloadServiceMock, LoginRepositoryMock, MockSessionPersister) and comprehensive LoginServiceTests to improve test coverage. Updated ConfigMockData, ConfigServiceMock, and UpdateServiceTests to include the lastBuildSize property for more accurate configuration mocking.

abbydomi committed:
feat: Move deviceAvailableSpace to App and update usage

Refactored deviceAvailableSpace from LoginService to App, making it part of the AppProtocol. Updated LoginService to use app.deviceAvailableSpace() and removed the local implementation. This centralizes device storage checks and improves code organization.

abbydomi committed:
feat: Add deviceAvailableSpace stub to AppMock

Introduced a stub property and method for deviceAvailableSpace in AppMock to allow tests to simulate available device storage. This supports testing features that depend on device storage availability.

abbydomi committed:
feat: Improve assertions and test setup in LoginServiceTests

Refines test assertions to check result type and error explicitly instead of matching the entire result. Updates test setup by stubbing device available space and open URL result where necessary to ensure accurate test conditions.

EduardoParadaApplivery commented on AppliverySDK/Wrappers/App.swift:
```diff
@@ -40,81 +39,90 @@ extension AppProtocol {
-		}
-		
-		return language
-	}
-	
-	func openUrl(_ urlString: String) -> Bool {
-		logInfo("Opening \(urlString)")
-		guard let url = URL(string: urlString) else { return false }
-		UIApplication.shared.open(url)
-		
-		return true
-	}
-	
+    }
+
+    func getLanguage() -> String {
+        guard let language = (Locale.current as NSLocale).object(forKey: NSLocale.Key.languageCode) as? String else {
+            return "NO_LANGUAGE_FOUND"
+        }
+
+        return language
+    }
+
+    func openUrl(_ urlString: String) -> Bool {
+        logInfo("Opening \(urlString)")
+        guard let url = URL(string: urlString) else { return false }
+        UIApplication.shared.open(url)
+
+        return true
+    }
+
+
```
It is recommended that there be only one line break between functions.

alexkater replied:
Must to have: Swiftlint Checks on here, WDYT @EduardoParadaApplivery. 

If there is no task, can you create it and take care of it?

Add to the project + add the check on the CI/CD

FranAlarza commented on AppliverySDK/Data/Persisters/SessionPersister.swift:
```diff
@@ -18,8 +18,14 @@ enum KeychainError: Error {
+struct SessionPersister: SessionPersisterProtocol {
 	let userDefaults: UserDefaultsProtocol
 	
 //	func save(accessToken: AccessToken?) {
```
If this function is commented out and not used, remove it. :)

FranAlarza commented on AppliverySDK/Wrappers/App.swift:
```diff
@@ -40,81 +39,90 @@ extension AppProtocol {
+    private lazy var alertPostpone: UIAlertController = UIAlertController()
+    private lazy var alertError: UIAlertController = UIAlertController()
+    private lazy var alertLogin: UIAlertController = UIAlertController()
+
```
Remove space

FranAlarza commented on AppliverySDK/Services/LoginService.swift:
```diff
@@ -123,7 +123,18 @@ final class LoginService: LoginServiceProtocol {
+            let freeSpace = try app.deviceAvailableSpace()
+            if freeSpace < (lastConfig.config?.lastBuildSize ?? 0) {
+                onResult?(.failure(error: .noDiskSpaceAvailable))
+                app.showErrorAlert("Insufficient storage")
```
Just out of curiosity, these show errors are being called in the correct thread, right? If they are not called in the main thread, the app will crash.

alexkater replied:
Why are we showing an alert from the Service layer? @FranAlarza this is common on the SDK?

FranAlarza reviewed (commented):
only minor changes and some doubts, good job!

alexkater reviewed (commented):
Good job! just some comments/questions

abbydomi committed:
style: remove unnecessary blank lines in App.swift

Cleaned up extra blank lines in the App class implementation to improve code readability and maintain consistency.

abbydomi committed:
style: remove commented-out save method in SessionPersister

Cleaned up SessionPersister by deleting an unused, commented-out save method to improve code readability.

abbydomi committed:
refactor: SDKData mock usage in tests

Added a static mock property to SDKData and updated LoginServiceTests to use SDKData.mock instead of duplicating mock initialization. This improves test maintainability and reduces code duplication.

abbydomi commented on AppliveryBehaviorTests/ServicesTests/LoginServiceTests.swift:
```diff
@@ -0,0 +1,266 @@
+        let forceUpdateConfig = UpdateConfigResponse(
+            config: SDKData(
+                minVersion: "1.0.0",
+                forceUpdate: true,
+                lastBuildId: "some-build-id",
+                mustUpdateMsg: nil,
+                ota: false,
+                lastBuildVersion: nil,
+                updateMsg: nil,
+                forceAuth: false,
+                lastBuildSize: 5000000000
+            ),
+            version: "0.9.0",
+            buildNumber: "100"
+        )
+        configService.currentConfigResponse = forceUpdateConfig
+        let loginService = LoginService(
+            loginRepository: loginRepository,
+            configService: configService,
+            downloadService: downloadService,
+            globalConfig: globalConfig,
+            sessionPersister: sessionPersister,
+            app: app,
+            keychain: keychain
+        )
```
I've updated it to have a static SDKData.mock 💪 Thanks for the suggestion

abbydomi committed:
refactor: Show error alerts on main thread in LoginService

Wrapped showErrorAlert calls in DispatchQueue.main.async to ensure UI updates occur on the main thread when handling storage errors during login.

abbydomi committed:
Merge branch 'develop' into feature/apple-51-improve-error-feedback-when-app-install-fails-due-to
